From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pulsar Bot <pulsar@localhost>
Date: Sat, 08 Aug 2026 12:00:00 +0000
Subject: [PATCH 2/2] futex: implement futex_waitv syscall (nr 449)

Backport of sys_futex_waitv from Linux 5.16 adapted for the 4.19 CAF
single-file futex implementation (kernel/futex.c).

futex_waitv allows waiting on an array of up to 128 futexes in a single
syscall. Critical for Wine/ntsync: games batch waits on multiple sync
objects instead of N individual futex syscalls.

The implementation is appended to kernel/futex.c reusing existing
infrastructure: futex_wait_setup(), unqueue_me(), hrtimer_sleeper.

Signed-off-by: Pulsar Bot <pulsar@localhost>
---
 arch/arm64/include/asm/unistd32.h |   4 +
 include/linux/syscalls.h          |   4 +
 include/uapi/asm-generic/unistd.h |   4 +
 kernel/futex.c                    | 172 ++++++++++++++++++++++++++++++
 kernel/sys_ni.c                   |   1 +
 5 files changed, 185 insertions(+)

diff --git a/arch/arm64/include/asm/unistd32.h b/arch/arm64/include/asm/unistd32.h
index aabbcc1..aabbcc2 100644
--- a/arch/arm64/include/asm/unistd32.h
+++ b/arch/arm64/include/asm/unistd32.h
@@ -834,4 +834,8 @@
 #define __NR_faccessat2 439
 __SYSCALL(__NR_faccessat2, sys_faccessat2)
+
+#define __NR_futex_waitv 449
+__SYSCALL(__NR_futex_waitv, sys_futex_waitv)
+
diff --git a/include/linux/syscalls.h b/include/linux/syscalls.h
index aabbcc3..aabbcc4 100644
--- a/include/linux/syscalls.h
+++ b/include/linux/syscalls.h
@@ -879,4 +879,8 @@
 asmlinkage long sys_sendmmsg(int fd, struct mmsghdr __user *msg,
 			     unsigned int vlen, unsigned flags);
+
+struct futex_waitv;
+asmlinkage long sys_futex_waitv(struct futex_waitv __user *waiters,
+				unsigned int nr_futexes, unsigned int flags,
+				struct __kernel_timespec __user *timeout,
+				clockid_t clockid);
+
diff --git a/include/uapi/asm-generic/unistd.h b/include/uapi/asm-generic/unistd.h
index aabbcc5..aabbcc6 100644
--- a/include/uapi/asm-generic/unistd.h
+++ b/include/uapi/asm-generic/unistd.h
@@ -758,4 +758,8 @@
 #define __NR_faccessat2 439
 __SYSCALL(__NR_faccessat2, sys_faccessat2)
+
+#define __NR_futex_waitv 449
+__SYSCALL(__NR_futex_waitv, sys_futex_waitv)
+
diff --git a/kernel/futex.c b/kernel/futex.c
index aabbcc7..aabbcc8 100644
--- a/kernel/futex.c
+++ b/kernel/futex.c
@@ -4195,3 +4195,175 @@ core_initcall(futex_init);
+
+/* ---------------------------------------------------------------
+ * futex_waitv: wait on multiple futexes (backport from Linux 5.16)
+ * --------------------------------------------------------------- */
+
+struct futex_waitv_q {
+	struct futex_q            q;
+	u32 __user               *uaddr;
+	u32                       val;
+	unsigned int              flags;
+	struct futex_hash_bucket *hb;
+};
+
+static int futex_waitv_setup_all(struct futex_waitv_q *qv, unsigned int count)
+{
+	unsigned int i;
+	int ret;
+
+	for (i = 0; i < count; i++) {
+		unsigned int kflags = 0;
+
+		if (!(qv[i].flags & FUTEX_PRIVATE_FLAG))
+			kflags |= FLAGS_SHARED;
+
+		qv[i].q = futex_q_init;
+
+		ret = futex_wait_setup(qv[i].uaddr, qv[i].val, kflags,
+				       &qv[i].q, &qv[i].hb);
+		if (ret) {
+			int j;
+
+			for (j = (int)i - 1; j >= 0; j--)
+				unqueue_me(&qv[j].q);
+			return ret;
+		}
+	}
+	return 0;
+}
+
+static int futex_waitv_dequeue_all(struct futex_waitv_q *qv, unsigned int count)
+{
+	unsigned int i;
+	int woken = -1;
+
+	for (i = 0; i < count; i++) {
+		if (!unqueue_me(&qv[i].q) && woken < 0)
+			woken = (int)i;
+	}
+	return woken;
+}
+
+SYSCALL_DEFINE5(futex_waitv,
+		struct futex_waitv __user *, uwaitv,
+		unsigned int, nr_futexes,
+		unsigned int, flags,
+		struct __kernel_timespec __user *, timeout,
+		clockid_t, clockid)
+{
+	struct hrtimer_sleeper to;
+	bool has_timeout = false;
+	struct futex_waitv_q *qv;
+	struct futex_waitv kv[FUTEX_WAITV_MAX];
+	unsigned int i;
+	int ret;
+
+	if (flags & ~FUTEX_WAITV_CLOCK_REALTIME)
+		return -EINVAL;
+	if (!nr_futexes || nr_futexes > FUTEX_WAITV_MAX)
+		return -EINVAL;
+
+	if (copy_from_user(kv, uwaitv,
+			   sizeof(struct futex_waitv) * nr_futexes))
+		return -EFAULT;
+
+	for (i = 0; i < nr_futexes; i++) {
+		if (kv[i].__reserved)
+			return -EINVAL;
+		if (!(kv[i].flags & FUTEX_32))
+			return -EINVAL;
+		if (kv[i].flags & ~(FUTEX_32 | FUTEX_PRIVATE_FLAG))
+			return -EINVAL;
+	}
+
+	qv = kcalloc(nr_futexes, sizeof(*qv), GFP_KERNEL);
+	if (!qv)
+		return -ENOMEM;
+
+	for (i = 0; i < nr_futexes; i++) {
+		qv[i].uaddr = (u32 __user *)(uintptr_t)kv[i].uaddr;
+		qv[i].val   = (u32)kv[i].val;
+		qv[i].flags = kv[i].flags;
+	}
+
+	if (timeout) {
+		struct timespec64 ts;
+		clockid_t clk;
+
+		if (get_timespec64(&ts, timeout)) {
+			ret = -EFAULT;
+			goto out_free;
+		}
+		if (!timespec64_valid(&ts)) {
+			ret = -EINVAL;
+			goto out_free;
+		}
+
+		clk = (flags & FUTEX_WAITV_CLOCK_REALTIME) ?
+		      CLOCK_REALTIME : CLOCK_MONOTONIC;
+
+		hrtimer_init_sleeper_on_stack(&to, clk, HRTIMER_MODE_ABS);
+		hrtimer_set_expires(&to.timer, timespec64_to_ktime(ts));
+		has_timeout = true;
+	}
+
+	ret = futex_waitv_setup_all(qv, nr_futexes);
+	if (ret)
+		goto out_timer;
+
+	set_current_state(TASK_INTERRUPTIBLE);
+
+	if (has_timeout)
+		hrtimer_start_expires(&to.timer, HRTIMER_MODE_ABS);
+
+	if (signal_pending(current)) {
+		__set_current_state(TASK_RUNNING);
+		futex_waitv_dequeue_all(qv, nr_futexes);
+		ret = -ERESTARTSYS;
+		goto out_timer;
+	}
+
+	freezable_schedule();
+	__set_current_state(TASK_RUNNING);
+
+	ret = futex_waitv_dequeue_all(qv, nr_futexes);
+	if (ret >= 0)
+		goto out_timer;
+
+	if (has_timeout && !to.task)
+		ret = -ETIMEDOUT;
+	else if (signal_pending(current))
+		ret = -ERESTARTSYS;
+	else
+		ret = -EAGAIN;
+
+out_timer:
+	if (has_timeout) {
+		hrtimer_cancel(&to.timer);
+		destroy_hrtimer_on_stack(&to.timer);
+	}
+out_free:
+	kfree(qv);
+	return ret;
+}
diff --git a/kernel/sys_ni.c b/kernel/sys_ni.c
index aabbcc9..aabbcca 100644
--- a/kernel/sys_ni.c
+++ b/kernel/sys_ni.c
@@ -298,3 +298,4 @@
 COND_SYSCALL(openat2);
 COND_SYSCALL(pidfd_getfd);
+COND_SYSCALL(futex_waitv);
